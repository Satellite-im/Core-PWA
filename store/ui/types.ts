import { TranslateResult } from 'vue-i18n'
import { FileMessage } from '~/types/textile/mailbox'
import { Glyph } from '~/types/ui/glyph'
import { Channel } from '~/types/ui/server'
import { Alert } from '~/libraries/ui/Alerts'
import { ConversationMessage } from '~/libraries/Iridium/chat/types'

export enum GlyphMarketViewStatus {
  HOME = 'home',
  SHOP_ALL = 'shop_all',
  SHOP_DETAIL = 'shop_detail',
}

export enum ModalWindows {
  NEW_FOLDER = 'newfolder',
  CREATE_SERVER = 'createServer',
  MARKETPLACE = 'marketplace',
  WALLET = 'wallet',
  WALLET_MINI = 'walletMini',
  ERROR = 'error',
  CHANGELOG = 'changelog',
  GLYPH = 'glyph',
  USER_PROFILE = 'userProfile',
  CALL_TO_ACTION = 'callToAction',
  RENAME_FILE = 'renameFile',
}

export interface EnhancerInfo {
  show: boolean
  floating?: boolean
  position?: number[]
  defaultWidth?: string
  defaultHeight?: string
  containerWidth?: number
  route: string
}

export interface EmojiUsage {
  code: string
  count: number
  content: string
}

export interface RecentGlyph {
  pack: Glyph
  url: string
  count: number
}

export enum SettingsRoutes {
  PERSONALIZE = 'personalize',
  PROFILE = 'profile',
  AUDIO_AND_VIDEO = 'audio & Video',
  KEY_BINDS = 'keybinds',
  ACCOUNTS_AND_DEVICES = 'accounts & Devices',
  PRIVACY = 'privacy',
  DEVELOPER = 'developer',
  INFO = 'info',
  NOTIFICATIONS = 'notifications',
  STORAGE = 'storage',
  NETWORK = 'network',
  REALMS = 'realms',
}

export type ContextMenuItemTypes = 'primary' | 'danger'

export interface ContextMenuItem {
  text: string | TranslateResult
  func: Function
  type?: ContextMenuItemTypes
}

export interface UIState {
  contextMenuStatus: boolean
  contextMenuValues: ContextMenuItem[]
  quickProfile: object | boolean
  userProfile: object
  notifications: Alert[]
  contextMenuPosition: object
  quickProfilePosition: object
  showSettings: boolean
  settingsSideBar: boolean
  settingsRoute: SettingsRoutes
  showSidebarUsers: boolean
  showSearchResult: boolean
  showSidebar: boolean
  modals: {
    [key in ModalWindows]: boolean | object
  }
  glyphModalPack: string
  chatbarContent: string
  chatbarFocus: boolean
  replyChatbarMessage?: ConversationMessage
  showPinned: boolean
  fullscreen: boolean
  enhancers: EnhancerInfo
  messages: any[]
  unreadMessage: number
  isScrollOver: boolean
  showOlderMessagesInfo: boolean
  isTyping: object | boolean
  isReacted: boolean
  activeChannel: Channel | undefined
  settingReaction: object
  hoveredGlyphInfo: object | undefined
  glyphMarketplaceView: object
  editMessage: {
    id: string
    from: string
    payload: string
  }
  mostEmojiUsed: EmojiUsage[]
  recentGlyphs: RecentGlyph[]
  chatImageOverlay?: FileMessage
  isMobileNavVisible: boolean
}

export type Position = {
  x: number
  y: number
}
