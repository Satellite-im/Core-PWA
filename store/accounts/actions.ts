import { Keypair } from '@solana/web3.js'
import Vue from 'vue'
import Crypto from '~/libraries/Crypto/Crypto'
import ServerProgram from '~/libraries/Solana/ServerProgram/ServerProgram'
import SolanaManager from '~/libraries/Solana/SolanaManager/SolanaManager'
import { ActionsArguments, RootState } from '../store.types'
import {
  AccountsError,
  RegistrationStatus,
  UserRegistrationPayload,
} from './types'

export default {
  /**
   * @method setPin DocsTODO
   * @description
   * @param pin
   * @example
   */
  async setPin({ commit }: ActionsArguments, pin: string) {
    if (pin.length < 5) {
      throw new Error(AccountsError.PIN_TOO_SHORT)
    }

    const $Crypto: Crypto = Vue.prototype.$Crypto

    const pinHash = await $Crypto.hash(pin)

    // The cleartext version of the pin will not be
    // persisted
    commit('setPin', pin)
    commit('setPinHash', pinHash)
  },
  /**
   * @method unlock DocsTODO
   * @description
   * @param pin
   * @example
   */
  async unlock({ commit, state, dispatch }: ActionsArguments, pin: string) {
    const { pinHash, encryptedPhrase } = state.accounts

    if (pin.length < 5) {
      throw new Error(AccountsError.PIN_TOO_SHORT)
    }

    const $Crypto: Crypto = Vue.prototype.$Crypto

    const computedPinHash = await $Crypto.hash(pin)

    if (computedPinHash !== pinHash) {
      throw new Error(AccountsError.INVALID_PIN)
    }

    if (encryptedPhrase !== '') {
      const decryptedPhrase = await $Crypto.decryptWithPassword(
        encryptedPhrase,
        pin
      )

      await commit('setPhrase', decryptedPhrase)
    }

    commit('unlock', pin)
    dispatch('loadAccount')
  },
  /**
   * @method generateWallet DocsTODO
   * @description
   * @param
   * @example
   */
  async generateWallet({ commit, state }: ActionsArguments) {
    const { pin } = state.accounts

    if (!pin) {
      throw new Error(AccountsError.INVALID_PIN)
    }

    const $SolanaManager: SolanaManager = Vue.prototype.$SolanaManager
    const $Crypto: Crypto = Vue.prototype.$Crypto

    const solanaWallet = await $SolanaManager.createRandomKeypair()

    if (!solanaWallet.mnemonic) {
      throw new Error(AccountsError.UNABLE_TO_CREATE_MNEMONIC)
    }

    await commit('setPhrase', solanaWallet.mnemonic)

    const encryptedPhrase = await $Crypto.encryptWithPassword(
      solanaWallet.mnemonic,
      pin
    )

    commit('setEncryptedPhrase', encryptedPhrase)
  },
  /**
   * @method loadAccount DocsTODO
   * @description
   * @param
   * @example
   */
  async loadAccount({ commit, state, dispatch }: ActionsArguments) {
    const $SolanaManager: SolanaManager = Vue.prototype.$SolanaManager

    const mnemonic = state.accounts.phrase

    if (mnemonic === '') {
      throw new Error(AccountsError.MNEMONIC_NOT_PRESENT)
    }

    await $SolanaManager.initializeFromMnemonic(mnemonic)

    const userAccount = await $SolanaManager.getUserAccount()

    if (!userAccount) {
      throw new Error(AccountsError.USER_DERIVATION_FAILED)
    }

    commit('setActiveAccount', userAccount?.publicKey.toBase58())

    const serverProgram: ServerProgram = new ServerProgram($SolanaManager)

    const userInfo = await serverProgram.getUser(userAccount.publicKey)

    if (userInfo === null) {
      throw new Error(AccountsError.USER_NOT_REGISTERED)
    }

    // Initialize Encryption Engine
    dispatch('initializeEncryptionEngine', userAccount)
    commit('setUserDetails', {
      username: userInfo.name,
      ...userInfo,
    })
  },
  /**
   * @method registerUser DocsTODO
   * @description
   * @param userData
   * @example
   */
  async registerUser(
    { commit, dispatch }: ActionsArguments,
    userData: UserRegistrationPayload
  ) {
    const $SolanaManager: SolanaManager = Vue.prototype.$SolanaManager

    commit('setRegistrationStatus', RegistrationStatus.IN_PROGRESS)

    const balance = await $SolanaManager.getCurrentAccountBalance()

    if (balance === 0) {
      commit('setRegistrationStatus', RegistrationStatus.FUNDING_ACCOUNT)
      await $SolanaManager.requestAirdrop()
    }

    const payerAccount = await $SolanaManager.getActiveAccount()

    if (!payerAccount) {
      commit('setRegistrationStatus', RegistrationStatus.UKNOWN)
      throw new Error(AccountsError.PAYER_NOT_PRESENT)
    }

    const userAccount = await $SolanaManager.getUserAccount()

    if (!userAccount) {
      commit('setRegistrationStatus', RegistrationStatus.UKNOWN)
      throw new Error(AccountsError.USER_DERIVATION_FAILED)
    }

    const serverProgram: ServerProgram = new ServerProgram($SolanaManager)

    const userInfo = await serverProgram.getUser(userAccount.publicKey)

    if (userInfo) {
      commit('setRegistrationStatus', RegistrationStatus.REGISTERED)
      throw new Error(AccountsError.USER_ALREADY_REGISTERED)
    }

    commit('setRegistrationStatus', RegistrationStatus.SENDING_TRANSACTION)

    await serverProgram.createUser(userData.name, '', userData.status)

    commit('setRegistrationStatus', RegistrationStatus.REGISTERED)

    commit('setActiveAccount', userAccount.publicKey.toBase58())

    // Initialize Encryption Engine
    dispatch('initializeEncryptionEngine', userAccount)
    commit('setUserDetails', {
      username: userData.name,
      status: userData.status,
      photoHash: userData.photoHash,
      address: userAccount.publicKey.toBase58(),
    })
  },
  /**
   * @method initializeEncryptionEngine DocsTODO
   * @description
   * @param userAccount
   * @example
   */
  async initializeEncryptionEngine(_: RootState, userAccount: Keypair) {
    // Initialize crypto engine
    const $Crypto: Crypto = Vue.prototype.$Crypto
    await $Crypto.init(userAccount)
  },
}
