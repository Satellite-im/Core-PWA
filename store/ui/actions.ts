import Mousetrap from 'mousetrap'
import Vue from 'vue'
import { v4 as uuidv4 } from 'uuid'
import { Position, UIState } from './types'
import SoundManager, { Sounds } from '~/libraries/SoundManager/SoundManager'
import TextileManager from '~/libraries/Textile/TextileManager'
import { ActionsArguments } from '~/types/store/store'
import { Friend } from '~/types/ui/friends'
import { Channel } from '~/types/ui/server'
import { getFullUserInfoFromState } from '~/utilities/Messaging'
import { getCorrectKeybind } from '~/utilities/Keybinds'
import { TextileError } from '~/store/textile/types'
import { AlertType } from '~/libraries/ui/Alerts'

const $Sounds = new SoundManager()

export default {
  setMessages({ commit }: ActionsArguments<UIState>, messages: any[]) {
    commit('setMessages', messages)
  },
  sendMessage({ commit, rootState }: ActionsArguments<UIState>, message: any) {
    if (message.user.address !== rootState.accounts.active) {
      $Sounds.playSound(Sounds.NEW_MESSAGE)
    }
    commit('sendMessage', message)
  },
  setIsScrollOver({ commit }: ActionsArguments<UIState>, status: boolean) {
    commit('setIsScrollOver', status)
  },
  setIsReacted({ commit }: ActionsArguments<UIState>, status: boolean) {
    commit('setIsReacted', status)
  },
  setActiveChannel({ commit }: ActionsArguments<UIState>, channel: Channel) {
    commit('setActiveChannel', channel)
  },
  /**
   * @method openSettings
   * @description Opens setting page
   * @example Mousetrap.bind('ctrl+s', dispatch('audio/toggleMute') )
   */
  openSettings({ commit, state }: any) {
    commit('toggleSettings', { show: !state.showSettings })
  },
  /**
   * @method activateKeybinds
   * @description Activates all keybindings with Mousetrap
   * @example mounted (){ activateKeybinds() }
   */
  async activateKeybinds({ dispatch, rootState }: ActionsArguments<UIState>) {
    const { toggleMute, toggleDeafen, openSettings, callActiveChat } =
      // @ts-ignore
      rootState.settings.keybinds
    Mousetrap.reset()
    Mousetrap.bind(getCorrectKeybind(toggleMute), (event: KeyboardEvent) => {
      event.preventDefault()
      dispatch('audio/toggleMute', null, { root: true })
    })
    Mousetrap.bind(getCorrectKeybind(toggleDeafen), (event: KeyboardEvent) => {
      event.preventDefault()
      dispatch('audio/toggleDeafen', null, { root: true })
    })
    Mousetrap.bind(getCorrectKeybind(openSettings), (event: KeyboardEvent) => {
      event.preventDefault()
      dispatch('openSettings')
    })
    Mousetrap.bind(
      getCorrectKeybind(callActiveChat),
      (event: KeyboardEvent) => {
        event.preventDefault()
        dispatch('webrtc/call', { kinds: ['audio'] }, { root: true })
      },
    )
  },
  /**
   * @method setNotifications
   * @description Collects all existing notifications for a user
   */
  async setNotifications({ commit }: ActionsArguments<UIState>) {
    const $TextileManager: TextileManager = Vue.prototype.$TextileManager

    if (!$TextileManager.notificationManager?.isInitialized()) {
      throw new Error(TextileError.MAILBOX_MANAGER_NOT_INITIALIZED)
    }
    const notifications =
      await $TextileManager.notificationManager?.getnotifications()
    commit('setNotifications', notifications)
  },
  async sendNotification(
    { commit, rootState }: ActionsArguments<UIState>,
    payload: {
      message: string
      from: string
      imageHash: string
      title: string
      type: AlertType
    },
  ) {
    const $TextileManager: TextileManager = Vue.prototype.$TextileManager

    if (!$TextileManager.notificationManager?.isInitialized()) {
      throw new Error(TextileError.MAILBOX_MANAGER_NOT_INITIALIZED)
    }
    const notificationResponse =
      await $TextileManager.notificationManager?.sendNotification({
        from: payload.from,
        id: uuidv4(),
        title: payload.title,
        imageHash: payload.imageHash,
        message: payload.message,
        type: payload.type,
      })
    commit('sendNotification', notificationResponse)
  },
  removeSeenNotification(
    { commit, rootState }: ActionsArguments<UIState>,
    notificationId: string,
  ) {
    commit('notificationSeen', notificationId)
  },
  /**
   * @method clearKeybinds
   * @description Unbinds all current keybindings with Mousetrap
   * @example destroyed (){ clearKeybinds() }
   */
  async clearKeybinds({ dispatch }: ActionsArguments<UIState>) {
    Mousetrap.reset()
  },
  async setChatbarFocus({ dispatch }: ActionsArguments<UIState>) {
    await dispatch('toggleChatbarFocus', false)
    dispatch('toggleChatbarFocus', true)
  },
  toggleChatbarFocus({ commit }: ActionsArguments<UIState>, flag: boolean) {
    commit('setChatbarFocus', flag)
  },
  setChatbarContent(
    { commit, dispatch }: ActionsArguments<UIState>,
    val: {
      content: string
      userId?: string
    },
  ) {
    commit('chatbarContent', val.content)
    if (val.userId)
      dispatch(
        'chat/setChatText',
        { value: val.content, userId: val.userId },
        { root: true },
      )
  },
  showQuickProfile(
    { commit, state, rootState }: ActionsArguments<UIState>,
    payload: { textilePublicKey: string; position: Position },
  ) {
    if (!payload?.textilePublicKey || !payload?.position) {
      return
    }

    const selectedUser = getFullUserInfoFromState(
      payload.textilePublicKey,
      rootState,
    )

    commit('setQuickProfilePosition', payload.position)
    commit('quickProfile', selectedUser)
  },
  async showProfile(
    { commit, rootState, dispatch }: ActionsArguments<UIState>,
    user: Friend,
  ) {
    if (!user) {
      return
    }
    let metadata = null
    if (!user.metadata) {
      const $TextileManager: TextileManager = Vue.prototype.$TextileManager
      metadata = await $TextileManager.metadataManager?.getFriendMetadata(
        user.address,
      )
    }
    commit('toggleModal', {
      name: 'userProfile',
      state: true,
    })
    const friend = {
      ...user,
    }
    if (metadata) {
      friend.metadata = metadata
      commit('friends/updateFriend', friend, { root: true })
    }
    commit('setUserProfile', friend)
  },
}
