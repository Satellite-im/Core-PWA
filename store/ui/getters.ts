import { EmojiUsage, UIState } from './types'
import { Alert, AlertState } from '~/libraries/ui/Alerts'

const getters = {
  getSortedMostUsedEmojis: (state: UIState): EmojiUsage[] => {
    return [...state.mostEmojiUsed].sort((a, b) => b.count - a.count)
  },
  getSortedRecentGlyphs: (state: UIState) => {
    return [...state.recentGlyphs].sort((a, b) => b.count - a.count)
  },
  showSidebar: (state: UIState) => {
    return state.showSidebar
  },
  allUnseenNotifications: (state: UIState): Alert[] => {
    return [...state.notifications]
      .sort((a, b) => {
        return b.at - a.at
      })
      .filter((noti) => {
        return noti.state === AlertState.UNREAD
      })
  },
  getAlertCount: (state: UIState) => {
    return state.notificationsAlertCounter
  },
  swiperSlideIndex: (state: UIState) => {
    return state.swiperSlideIndex
  },
  isFilesIndexLoading: (state: UIState): boolean => {
    return Boolean(state.filesUploadStatus)
  },
}

export default getters
