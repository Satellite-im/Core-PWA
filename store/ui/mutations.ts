import { NuxtState } from '@nuxt/types/app'

export default {
  toggleContextMenu(state: NuxtState, enabled: Boolean) {
    state.ui = {
      ...state.ui,
      contextMenuStatus: enabled,
    }
  },
  showSidebarUsers(state: NuxtState, enabled: Boolean) {
    state.ui = {
      ...state.ui,
      showSidebarUsers: enabled,
    }
  },
  setContextMenuValues(state: NuxtState, values: any) {
    state.ui = {
      ...state.ui,
      contextMenuValues: values,
    }
  },
  setContextMenuPosition(state: NuxtState, e: any) {
    state.ui = {
      ...state.ui,
      contextMenuPosition: { x: e.x, y: e.y },
    }
  },
  setQuickProfilePosition(state: NuxtState, e: any) {
    state.ui = {
      ...state.ui,
      quickProfilePosition: { x: e.x, y: e.y },
    }
  },
  quickProfile(state: NuxtState, profile: Object | Boolean) {
    state.ui = {
      ...state.ui,
      quickProfile: profile,
    }
  },
  chatbarContent(state: NuxtState, content: String) {
    state.ui = {
      ...state.ui,
      chatbarContent: content,
    }
  },
  fullscreen(state: NuxtState, fullscreen: Boolean) {
    state.ui = {
      ...state.ui,
      fullscreen,
    }
  },
  toggleEnhancers(state: NuxtState, show: Boolean) {
    state.ui = {
      ...state.ui,
      showEnhancers: show,
    }
  },
}
