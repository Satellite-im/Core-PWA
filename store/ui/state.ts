interface UIState {
  contextMenuStatus: Boolean
  contextMenuValues: Array<Object>
  quickProfile: Object | Boolean
  contextMenuPosition: Object
  quickProfilePosition: Object
  showSettings: Boolean
  showSidebarUsers: Boolean
  showSearchResult: Boolean
  modals: Object
  chatbarContent: String
  replyChatbarContent: {
    id: String
    from: String
    payload: String
  }
  fullscreen: Boolean
  showEnhancers: Boolean
  messages: any[]
  unreadMessage: number
  isScrollOver: boolean
  isTyping: Object | Boolean
}

const InitalUIState = (): UIState => ({
  contextMenuStatus: false,
  showSidebarUsers: true,
  showSearchResult: false,
  showSettings: false,
  quickProfile: false,
  contextMenuValues: [],
  contextMenuPosition: { x: 0, y: 0 },
  quickProfilePosition: { x: 0, y: 0 },
  modals: {
    newfolder: false,
    createServer: false,
  },
  chatbarContent: '',
  replyChatbarContent: { id: '', from: '', payload: '' },
  fullscreen: false,
  showEnhancers: false,
  messages: [],
  unreadMessage: 0,
  isScrollOver: false,
  isTyping: false,
})

export default InitalUIState
