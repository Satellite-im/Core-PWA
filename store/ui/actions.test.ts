import Vue from 'vue'
import Mousetrap from 'mousetrap'
import * as actions from '~/store/ui/actions'
import { RegistrationStatus } from '~/store/accounts/types'
import { DataStateType } from '~/store/dataState/types'
import { CaptureMouseTypes } from '~/store/settings/types'
import SoundManager from '~/libraries/SoundManager/SoundManager'
import TextileManager from '~/libraries/Textile/TextileManager'
Vue.prototype.$TextileManager = new TextileManager()

const $Sounds = new SoundManager()

const initialRootState: any = {
  accounts: {
    storePin: true,
    loading: true,
    locked: true,
    pin: '',
    pinHash: '',
    active: '',
    gasPrice: '',
    phrase: '',
    error: '',
    encryptedPhrase: '',
    registered: true,
    details: {
      name: '',
      address: '',
      status: '',
      state: 'idle',
      unreadCount: 123,
      profilePicture: '',
      badge: 'community',
      userAccount: '',
      mailboxId: '',
      textilePubkey: '',
    },
    registrationStatus: RegistrationStatus.IN_PROGRESS,
    lastVisited: '',
  },
  dataState: {
    files: DataStateType.Empty,
    friends: DataStateType.Loading,
    search: DataStateType.Ready,
  },
  friends: {
    incomingRequests: [
      {
        requestId: '',
        account: {
          accountId: '',
          from: '',
          status: 123,
          fromMailboxId: '',
          toMailboxId: '',
          to: '',
        },
        pending: true,
        from: '',
        userInfo: {
          name: '',
          servers: {},
          status: '',
          photoHash: '',
        },
      },
    ],
    outgoingRequests: [
      {
        to: '',
        requestId: '',
        account: {
          accountId: '',
          from: '',
          status: 123,
          fromMailboxId: '',
          toMailboxId: '',
          to: '',
        },
        pending: true,
      },
    ],
    all: [
      {
        publicKey: 'NoWiFi4you',
        typingState: 'NOT_TYPING',
        item: {},
        pending: true,
        activeChat: true,
        encryptedTextilePubkey: '',
        name: 'Taurus Nix',
        address: '0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f',
        account: {
          accountId: 'Checking Account',
          from: '.',
          status: 429,
          fromMailboxId: '12345',
          toMailboxId: 'v4.0.0-rc.4',
          to: './path/to/file',
        },
        textilePubkey: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
        status: '',
        state: 'idle',
        unreadCount: 123,
        profilePicture: '',
        badge: 'community',
        userAccount: '',
        mailboxId: '',
      },
    ],
  },
  textile: {
    initialized: true,
    conversations: {},
    conversationLoading: true,
    messageLoading: true,
    uploadProgress: {
      abc: {
        progress: 42,
        finished: false,
        name: 'file.pdf',
      },
    },
  },
  settings: {
    audioInput: '',
    audioOutput: '',
    videoInput: '',
    captureMouse: CaptureMouseTypes.always,
    noiseSuppression: true,
    echoCancellation: true,
    bitrate: 96000,
    sampleSize: 24,
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    userHasGivenAudioAccess: false,
    userDeniedAudioAccess: false,
    keybinds: {
      toggleMute: 'alt+m',
      toggleDeafen: 'alt+d',
      openSettings: 'alt+s',
      callActiveChat: 'alt+c',
    },
    embeddedLinks: true,
    displayCurrentActivity: true,
  },
}
const initialState = {
  contextMenuStatus: false,
  showSidebarUsers: true,
  showSidebar: true,
  showSearchResult: false,
  showSettings: false,
  showMedia: false,
  settingsSideBar: true,
  settingsRoute: 'personalize',
  quickProfile: false,
  userProfile: {},
  contextMenuValues: [],
  contextMenuPosition: {
    x: 0,
    y: 0,
  },
  quickProfilePosition: {
    x: 0,
    y: 0,
  },
  modals: {
    changelog: false,
    createServer: false,
    error: false,
    glyph: false,
    marketplace: false,
    newfolder: false,
    quickchat: false,
    userProfile: false,
    wallet: false,
    walletMini: false,
  },
  glyphModalPack: '',
  chatbarContent: '',
  replyChatbarContent: {
    from: '',
    id: '',
    payload: '',
  },
  chatbarFocus: false,
  fullscreen: false,
  showPinned: false,
  enhancers: {
    containerWidth: 0,
    defaultHeight: '30rem',
    defaultWidth: '24rem',
    floating: false,
    position: [0, 0],
    route: 'emotes',
    show: false,
  },
  messages: [],
  unreadMessage: 0,
  isScrollOver: false,
  isTyping: {
    address: '0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c',
    last_message:
      'Ship of the imagination consciousness across the centuries network of wormholes finite but unbounded inconspicuous motes of rock and gas.',
    name: 'Phoenix Kalindi',
    profilePicture: '',
    state: 'online',
    status: 'Working on the space station',
    unreads: 4,
  },
  isReacted: false,
  activeChannel: undefined,
  settingReaction: {
    groupID: null,
    messageID: null,
    status: false,
  },
  hoveredGlyphInfo: undefined,
  glyphMarketplaceView: {
    shopId: null,
    view: 'home',
  },
  editMessage: {
    from: '',
    id: '',
    payload: '',
  },
  recentReactions: ['😰', '😡', '👍'],
  mostEmojiUsed: [],
  recentGlyphs: [],
  theme: {
    base: {
      class: '',
      name: 'default',
      text: 'Default',
      value: 'default',
    },
    flair: {
      text: 'Satellite',
      value: '#2761fd',
    },
  },
}

describe('init', () => {
  test('setMessages', () => {
    const commit = jest.fn()
    const messages = ['message 1']
    actions.default.setMessages({ commit }, messages)
    expect(commit).toHaveBeenCalledWith('setMessages', messages)
  })
  test('sendMessage', () => {
    const commit = jest.fn()
    const rootState: any = {
      accounts: {
        storePin: true,
        loading: true,
        locked: true,
        pin: '',
        pinHash: '',
        active: 'aktief',
        gasPrice: '',
        phrase: '',
        error: '',
        encryptedPhrase: '',
        registered: true,
        details: {
          name: '',
          address: '',
          status: '',
          state: 'idle',
          unreadCount: 123,
          profilePicture: '',
          badge: 'community',
          userAccount: '',
          mailboxId: '',
          textilePubkey: '',
        },
        lastVisited: '',
      },
    }
    const message = {
      user: {
        address: 'aktief',
      },
    }

    actions.default.sendMessage({ commit, rootState }, message)
    expect(commit).toHaveBeenCalledWith('sendMessage', message)
  })
  test('sendMessage with non-matching addresses', () => {
    const commit = jest.fn()
    const rootState: any = {
      accounts: {
        storePin: true,
        loading: true,
        locked: true,
        pin: '',
        pinHash: '',
        active: 'aktief',
        gasPrice: '',
        phrase: '',
        error: '',
        encryptedPhrase: '',
        registered: true,
        details: {
          name: '',
          address: '',
          status: '',
          state: 'idle',
          unreadCount: 123,
          profilePicture: '',
          badge: 'community',
          userAccount: '',
          mailboxId: '',
          textilePubkey: '',
        },
        lastVisited: '',
      },
    }
    const message = {
      user: {
        address: 'not aktief',
      },
    }

    window.HTMLMediaElement.prototype.load = () => {
      return Promise.resolve()
    }
    window.HTMLMediaElement.prototype.play = () => {
      return Promise.resolve()
    }
    window.HTMLMediaElement.prototype.pause = () => {
      return Promise.resolve()
    }

    actions.default.sendMessage({ commit, rootState }, message)
    expect(commit).toHaveBeenCalledWith('sendMessage', message)
  })
  test('setIsScrollOver', () => {
    const commit = jest.fn()
    actions.default.setIsScrollOver({ commit }, false)
    expect(commit).toHaveBeenCalledWith('setIsScrollOver', false)
  })
  test('setIsReacted', () => {
    const commit = jest.fn()
    actions.default.setIsReacted({ commit }, false)
    expect(commit).toHaveBeenCalledWith('setIsReacted', false)
  })
  test('setActiveChannel', () => {
    const commit = jest.fn()
    actions.default.setActiveChannel(
      { commit },
      {
        type: 'type',
        id: 'id',
        name: 'name',
      },
    )
    expect(commit).toHaveBeenCalledWith('setActiveChannel', {
      type: 'type',
      id: 'id',
      name: 'name',
    })
  })
  test('addReaction', () => {
    const commit = jest.fn()
    actions.default.addReaction(
      { commit },
      {
        id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9',
        groupID: '9876',
        replyID: 'bc23a9d531064583ace8f67dad60f6bb',
        messageID: '9876',
        emoji: '😌',
        reactor: {
          reactions: [
            'c466a48309794261b64a4f02cfcc3d64',
            'bc23a9d531064583ace8f67dad60f6bb',
            'da7588892',
            '9876',
          ],
          reactors: [true, true, true, true],
        },
      },
    )
    expect(commit).toHaveBeenCalledWith('addReaction', {
      id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9',
      groupID: '9876',
      replyID: 'bc23a9d531064583ace8f67dad60f6bb',
      messageID: '9876',
      emoji: '😌',
      reactor: {
        reactions: [
          'c466a48309794261b64a4f02cfcc3d64',
          'bc23a9d531064583ace8f67dad60f6bb',
          'da7588892',
          '9876',
        ],
        reactors: [true, true, true, true],
      },
    })
    expect(commit).toHaveBeenCalledWith('updateRecentReactions', '😌')
  })
  test('openSettings', () => {
    const commit = jest.fn()
    const state = { ...initialState }
    actions.default.openSettings({ commit, state })
    expect(commit).toHaveBeenCalledWith('toggleSettings', {
      show: !state.showSettings,
    })
  })
  test('setChatbarFocus', async () => {
    const dispatch = jest.fn()
    await actions.default.setChatbarFocus({ dispatch })
    expect(dispatch).toHaveBeenCalledWith('toggleChatbarFocus', false)
    expect(dispatch).toHaveBeenCalledWith('toggleChatbarFocus', true)
  })
  test('toggleChatbarFocus', async () => {
    const commit = jest.fn()
    const flag = true

    actions.default.toggleChatbarFocus({ commit }, flag)
    expect(commit).toHaveBeenCalledWith('setChatbarFocus', flag)
  })
  test('activateKeybinds', async () => {
    const dispatch = jest.fn()
    const rootState = { ...initialRootState }
    Mousetrap.reset = jest.fn()
    Mousetrap.bind = jest.fn()
    await actions.default.activateKeybinds({ dispatch, rootState })
    expect(Mousetrap.reset).toHaveBeenCalled()
    expect(Mousetrap.bind).toHaveBeenCalled()
    expect(Mousetrap.bind).toBeCalledTimes(4)
  })
  test('clearKeybinds', async () => {
    const dispatch = jest.fn()
    Mousetrap.reset = jest.fn()
    await actions.default.clearKeybinds({ dispatch })
    expect(Mousetrap.reset).toHaveBeenCalled()
  })
  test('setChatbarContent without userId', async () => {
    const commit = jest.fn()
    const dispatch = jest.fn()
    const val = {
      content: 'content',
    }
    await actions.default.setChatbarContent({ commit, dispatch }, val)
    expect(commit).toHaveBeenCalledWith('chatbarContent', val.content)
  })
  test('setChatbarContent with userId', async () => {
    const commit = jest.fn()
    const dispatch = jest.fn()
    const val = {
      content: 'content',
      userId: '0x1',
    }
    await actions.default.setChatbarContent({ commit, dispatch }, val)
    expect(commit).toHaveBeenCalledWith('chatbarContent', val.content)
    expect(dispatch).toHaveBeenCalledWith(
      'chat/setChatText',
      { value: val.content, userId: val.userId },
      { root: true },
    )
  })
  test('showQuickProfile with payload', async () => {
    const commit = jest.fn()
    const rootState = { ...initialRootState }
    const localInitState = { ...initialState }
    const payload = {
      textilePublicKey: 'public key',
      position: {
        x: 6,
        y: 2,
      },
    }
    await actions.default.showQuickProfile(
      { commit, localInitState, rootState },
      payload,
    )
    expect(commit).toHaveBeenCalledWith(
      'setQuickProfilePosition',
      payload.position,
    )
  })
  test('showQuickProfile without payload', async () => {
    const commit = jest.fn()
    const localInitState = { ...initialState }
    const rootState = { ...initialRootState }
    const payload = {}
    const result = await actions.default.showQuickProfile(
      { commit, localInitState, rootState },
      payload,
    )
    expect(commit).not.toHaveBeenCalledWith(
      'setQuickProfilePosition',
      payload.position,
    )
    expect(result).toBeUndefined()
  })
  test('showProfile with existing friend metadata', async () => {
    const TMConstructor = Vue.prototype.$TextileManager
    TMConstructor.metadataManager = jest.fn()
    TMConstructor.metadataManager.getFriendMetadata = jest
      .fn()
      .mockReturnValueOnce({
        note: 'friend metadata',
      })

    const commit = jest.fn()
    const rootState = { ...initialRootState }
    const dispatch = jest.fn()

    await actions.default.showProfile(
      { commit, rootState, dispatch },
      rootState.friends,
    )
    expect(commit).toHaveBeenCalledWith('toggleModal', {
      name: 'userProfile',
      state: true,
    })
  })
  test('showProfile with no friend metadata', async () => {
    const TMConstructor = Vue.prototype.$TextileManager

    const commit = jest.fn()
    const rootState = { ...initialRootState }
    const dispatch = jest.fn()

    await actions.default.showProfile(
      { commit, rootState, dispatch },
      rootState.friends,
    )
    expect(commit).toHaveBeenCalledWith('toggleModal', {
      name: 'userProfile',
      state: true,
    })
  })
  test('showProfile with no passed-in user argument', async () => {
    const TMConstructor = Vue.prototype.$TextileManager

    const commit = jest.fn()
    const rootState = { ...initialRootState }
    const dispatch = jest.fn()

    const result = await actions.default.showProfile(
      { commit, rootState, dispatch },
      null,
    )
    expect(result).toBeUndefined()
  })
})
