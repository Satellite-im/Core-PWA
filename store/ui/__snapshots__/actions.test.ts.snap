// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions.default.showQuickProfile 0 1`] = `undefined`;

exports[`actions.default.showQuickProfile 1 1`] = `undefined`;

exports[`actions.default.showQuickProfile 2 1`] = `undefined`;

exports[`actions.default.showQuickProfile 3 1`] = `undefined`;

exports[`actions.default.showQuickProfile 4 1`] = `undefined`;

exports[`actions.default.showQuickProfile 5 1`] = `undefined`;
